# SQL DATA ANALYSIS

# Goals
For this project, we are to connect to a Postgresql database in order to run SQL queries on the following data from a startup: books, publishers, authors, and customer ratings and reviews of books. 

The resulting analysis will be used to generate a value proposition for a new product. 

# Tasks
    * Find the number of books released after January 1, 2000.
    * Find the number of user reviews and the average rating for each book.
    * Identify the publisher that has released the greatest number of books with more than 50 pages.
    * Identify the author with the highest average book rating: look only at books with at least 50 ratings.
    * Find the average number of text reviews among users who rated more than 50 books.

# Data Initialization

In [1]:
# import libraries
import pandas as pd
from sqlalchemy import create_engine

In [2]:
# connect to database
db_config = {'user': 'praktikum_student',         # username
             'pwd': 'Sdf4$2;d-d30pp', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # connection port
             'db': 'data-analyst-final-project-db'}          # the name of the database

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

## Basic SQL Queries

In [3]:
# top of books table
books_head = pd.io.sql.read_sql('SELECT * FROM books;', con = engine)
books_head.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [4]:
# check general info
books_head.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


In [5]:
# check for duplicates
books_head.duplicated().sum()

0

There are no missing values or duplicates in the **books** table.

In [6]:
# top of authors table
authors_head = pd.io.sql.read_sql('SELECT * FROM authors;', con = engine)
authors_head.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [7]:
# check general info
authors_head.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


In [8]:
# check for duplicates
authors_head.duplicated().sum()

0

There are no missing values or duplicates in the **authors** table.

In [9]:
# top of publishers table
publishers_head = pd.io.sql.read_sql('SELECT * FROM publishers;', con = engine)
publishers_head.head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [10]:
# check general info
publishers_head.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


In [11]:
# check for duplicates
publishers_head.duplicated().sum()

0

There are no missing values or duplicates in the **publishers** table.

In [12]:
# top of ratings table
ratings_head = pd.io.sql.read_sql('SELECT * FROM ratings;', con = engine)
ratings_head.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [13]:
# check general info
ratings_head.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


In [14]:
# check for duplicates
ratings_head.duplicated().sum()

0

There are no missing values or duplicates in the **ratings** table.

In [15]:
# top of reviews table
reviews_head = pd.io.sql.read_sql('SELECT * FROM reviews;', con = engine)
reviews_head.head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [16]:
# check general info
reviews_head.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


In [17]:
# check for duplicates
reviews_head.duplicated().sum()

0

There are no missing values or duplicates in the **reviews** table.

Overall, all tables have no missing values or duplicates due to their primary ID keys as shown above.

# Tasks

## Find the number of books released after January 1, 2000

In [18]:
# number of books after 2000-01-01
books_count = pd.io.sql.read_sql('''SELECT COUNT(DISTINCT book_id) AS book_cnt FROM books WHERE (publication_date :: date) > '2000-01-01';''', con = engine)
books_count

book_cnt
0       819

There are 819 books released after January 1, 2000.

## Find the number of user reviews and the average rating for each book

In [19]:
# books, review count and avergae rating value in order of review count
books_ratings_reviews_a = pd.io.sql.read_sql('SELECT title, SUBQ2.review_cnt, SUBQ2.rating_avg FROM books LEFT JOIN (SELECT SUBQ.book_id, COUNT(reviews.review_id) AS review_cnt, SUBQ.rating_avg FROM reviews RIGHT JOIN (SELECT book_id, AVG(ratings.rating) AS rating_avg FROM ratings GROUP BY book_id) AS SUBQ ON SUBQ.book_id = reviews.book_id GROUP BY SUBQ.book_id, SUBQ.rating_avg) AS SUBQ2 ON SUBQ2.book_id = books.book_id ORDER BY review_cnt DESC, rating_avg DESC;', con = engine)
books_ratings_reviews_a


title  review_cnt  rating_avg
0                              Twilight (Twilight  #1)           7    3.662500
1    Harry Potter and the Prisoner of Azkaban (Harr...           6    4.414634
2    Harry Potter and the Chamber of Secrets (Harry...           6    4.287500
3                                       The Book Thief           6    4.264151
4                                     The Glass Castle           6    4.206897
..                                                 ...         ...         ...
995  Disney's Beauty and the Beast (A Little Golden...           0    4.000000
996                               Leonardo's Notebooks           0    4.000000
997                          Essential Tales and Poems           0    4.000000
998    Anne Rice's The Vampire Lestat: A Graphic Novel           0    3.666667
999                            The Natural Way to Draw           0    3.000000

[1000 rows x 3 columns]

In [27]:
# books, review count and avergae rating value in order of rating average
books_ratings_reviews_b = pd.io.sql.read_sql('SELECT title, SUBQ2.review_cnt, SUBQ2.rating_avg FROM books LEFT JOIN (SELECT SUBQ.book_id, COUNT(reviews.review_id) AS review_cnt, SUBQ.rating_avg FROM reviews RIGHT JOIN (SELECT book_id, AVG(ratings.rating) AS rating_avg FROM ratings GROUP BY book_id) AS SUBQ ON SUBQ.book_id = reviews.book_id GROUP BY SUBQ.book_id, SUBQ.rating_avg) AS SUBQ2 ON SUBQ2.book_id = books.book_id ORDER BY rating_avg DESC, review_cnt DESC;', con = engine)
books_ratings_reviews_b


title  review_cnt  rating_avg
0                        A Dirty Job (Grim Reaper  #1)           4        5.00
1         Moneyball: The Art of Winning an Unfair Game           3        5.00
2              School's Out—Forever (Maximum Ride  #2)           3        5.00
3             Evil Under the Sun (Hercule Poirot  #24)           2        5.00
4                                            Geek Love           2        5.00
..                                                 ...         ...         ...
995  The World Is Flat: A Brief History of the Twen...           3        2.25
996                                      Drowning Ruth           3        2.00
997                  His Excellency: George Washington           2        2.00
998                                              Junky           2        2.00
999                               Harvesting the Heart           2        1.50

[1000 rows x 3 columns]

The book *A Dirty Job(Grim Reaper #1)* has a high rating average of 5.00 and a high review count of 4 while *Harvesting the Heart* has a low review count of 2 with the lowest rating average at 1.50. Per the first table, there are a few books with no reviews but relatively high ratings.

## Identify the publisher that has released the greatest number of books with more than 50 pages

In [21]:
# top publisher
top_publisher = pd.io.sql.read_sql('SELECT publisher FROM publishers, (SELECT publisher_id, COUNT(books.book_id) AS id_book FROM books WHERE num_pages > 50 GROUP BY publisher_id) AS SUBQ WHERE id_book = (SELECT MAX(id_book) FROM (SELECT publisher_id, COUNT(books.book_id) AS id_book FROM books WHERE num_pages > 50 GROUP BY publisher_id) AS SUBQ2 WHERE SUBQ.publisher_id = publishers.publisher_id);', con = engine)
top_publisher


publisher
0  Penguin Books

Based on our prompt above, the publisher with the greatest number of books with over 50 pages is Penguin Books with total count of 42 books recorded in the dataset.

## Identify the author with the highest average book rating: look only at books with at least 50 ratings

In [22]:
# author with the highest rating
high_rated_author = pd.io.sql.read_sql('SELECT author FROM authors,(SELECT books.book_id AS id_book, books.author_id AS id_author, COUNT(ratings.rating_id) AS rating_cnt, AVG(ratings.rating) AS rating_avg FROM books LEFT JOIN authors ON authors.author_id = books.author_id LEFT JOIN ratings ON ratings.book_id = books.book_id GROUP BY id_book HAVING COUNT(ratings.rating_id) >= 50) AS SUBQ WHERE rating_avg = (SELECT MAX(rating_avg) FROM (SELECT books.book_id AS id_book, books.author_id AS id_author, COUNT(ratings.rating_id) AS rating_cnt, AVG(ratings.rating) AS rating_avg FROM books LEFT JOIN authors ON authors.author_id = books.author_id LEFT JOIN ratings ON ratings.book_id = books.book_id GROUP BY id_book HAVING COUNT(ratings.rating_id) >= 50)AS SUBQ2 WHERE SUBQ.id_author = authors.author_id);', con = engine)
high_rated_author


author
0  J.K. Rowling/Mary GrandPré

The author with the highest rating average with over 50 ratings is *J.K. Rowling/Mary GrandPre* with an average of 4.41.

## Find the average number of text reviews among users who rated more than 50 books

In [23]:
# people who rated over 50 books and book count
count_ratings = pd.io.sql.read_sql('SELECT name, rating_cnt FROM (SELECT ratings.username AS name, COUNT(ratings.rating_id) AS rating_cnt FROM ratings  GROUP BY name HAVING COUNT(ratings.rating_id) > 50) AS SUBQ;', con = engine)
count_ratings


name  rating_cnt
0     sfitzgerald          55
1  jennifermiller          53
2          xdavis          51
3          paul88          56
4      martinadam          56
5       richard89          55

In [24]:
# number of reviews of people who rated over 50 books
count_reviews = pd.io.sql.read_sql('SELECT SUBQ.name AS name2, COUNT(reviews.review_id) AS review_cnt FROM reviews, (SELECT ratings.username AS name, COUNT(ratings.rating_id) AS rating_cnt FROM ratings  GROUP BY name HAVING COUNT(ratings.rating_id) > 50) AS SUBQ WHERE SUBQ.name = reviews.username GROUP BY name2;', con = engine)
count_reviews


name2  review_cnt
0          paul88          22
1          xdavis          18
2      martinadam          27
3       richard89          26
4     sfitzgerald          28
5  jennifermiller          25

In [25]:
# average number of reviews of people who rated over 50 books
avg_reviews = pd.io.sql.read_sql('SELECT AVG(review_cnt) FROM (SELECT SUBQ.name AS name2, COUNT(reviews.review_id) AS review_cnt FROM reviews, (SELECT ratings.username AS name, COUNT(ratings.rating_id) AS rating_cnt FROM ratings  GROUP BY name HAVING COUNT(ratings.rating_id) > 50) AS SUBQ WHERE SUBQ.name = reviews.username GROUP BY name2) AS SUBQ2;', con = engine)
avg_reviews


avg
0  24.333333

Based on the results of our query, only 6 individuals rated over 50 books with text review counts ranging from 18 to 28. Thus, the average number of their total review count is 24.3.

# General Conclusions

For all the tables provided in the database, there were no missing values or duplicates found largely due to the presence of primary keys (ID) on all the tables.

There were 819 books released after January 1, 2000 recorded in the database.

From our analysis, it could be seen that a high number of reviews or customer interest for the books does not equal a high rating. There were a few books with little to no reviews which had relatively high ratings. Of course, this doesn't mean that all people who leave reviews rate the books and vice versa.

Based on our query, the publisher with the greatest number of books with over 50 pages is Penguin Books with a total count of 42 books recorded in the database.

Also, the author with the highest rating average with over 50 ratings is J.K. Rowling/Mary GrandPre with an average of 4.41.

Finally, only 6 individuals rated more than 50 books with text review counts ranging from 18 to 28. The average number of their total review count is 24.3.

The database provided does not give ample information on the consumers, genre of books, regions represented or prices thus, we cannot make an informed decision on profitable new inclusions.